### PR TITLE
Add projects and components for validation into the config file

### LINKED ISSUE #30: Add projects and components for validation into the config file
Pseudo code: `Map<Project, List<Component>>` against which we will validate all the selected options in the config file for `participants`

### PR DESCRIPTION
**Pull request by xjusko**
resolves #30

### CODE REVIEW

File: src/main/java/org/jboss/config/LotteryConfig.java [hunk 1]
```diff
@@ -10,35 +10,64 @@
...
-                    @JsonProperty(required = true) int maxIssues) {
-                this(project, components, new Participation(maxIssues));
+    public void validate() throws ConfigException {
+        validateParticipantProjects();
+    }
+
+    private void validateParticipantProjects() throws ConfigException {
+        for (Participant participant : participants) {
+            for (Project project : participant.projects) {
+                if (!availableProjects.containsKey(project.project)) {
```
Comment (The-Huginn) [inline src/main/java/org/jboss/config/LotteryConfig.java:54]: Maybe we should do case-insensitive string comparison. I believe JQL is case insensitive as well

File: src/main/java/org/jboss/config/LotteryConfig.java [hunk 2]
```diff
@@ -10,35 +10,64 @@
...
+                    throw new ConfigException(
+                            "Participant " + participant.email + " has an invalid project: " + project.project);
+                }
+
+                Set<String> availableComponents = availableProjects.get(project.project);
+                if (!availableComponents.containsAll(project.components)) {
```
Comment (The-Huginn) [inline src/main/java/org/jboss/config/LotteryConfig.java:60]: same here

File: src/main/java/org/jboss/config/LotteryConfig.java [hunk 3]
```diff
@@ -10,35 +10,64 @@
...
+                    throw new ConfigException("Participant " + participant.email + " has invalid components for project "
+                            + project.project + ": " + project.components);
+                }
             }
         }
+    }
 
-        public record Participation(@JsonProperty(required = true) int maxIssues) {
+    public static class ConfigException extends Exception {
```
Comment (The-Huginn) [inline src/main/java/org/jboss/config/LotteryConfig.java:68]: I am not sure, if we need a custom exception here.

File: src/main/java/org/jboss/config/LotteryConfig.java [hunk 4]
```diff
@@ -10,35 +10,64 @@
...
-        public record Project(
-                @JsonProperty(required = true) String project,
-                @JsonDeserialize(as = TreeSet.class) Set<String> components,
-                @JsonUnwrapped @JsonProperty(access = JsonProperty.Access.READ_ONLY) Participation participation) {
-            // https://stackoverflow.com/a/71539100/6692043
-            @JsonCreator
-            public Project(@JsonProperty(required = true) String project, @JsonProperty(required = true) Set<String> components,
-                    @JsonProperty(required = true) int maxIssues) {
-                this(project, components, new Participation(maxIssues));
+    public void validate() throws ConfigException {
```
Comment (The-Huginn) [inline src/main/java/org/jboss/config/LotteryConfig.java:47]: I would probably keep the config just for config and validation in a different file or directly in LotteryConfigProducer could have this method as private.

File: src/main/resources/application.properties [hunk 5]
```diff
@@ -1,3 +1,3 @@
 camel.component.jira.jira-url=https://issues.redhat.com/
 %test.jira-issue-lottery.access-token=ignored
-jira-issue-lottery.config-file-repo=https://github.com/The-Huginn/jira-issue-lottery
```
Comment (The-Huginn) [inline src/main/resources/application.properties:3]: I am not sure about this, perhaps each of us should keep this in `.env` file and mention this in readme

File: src/test/java/org/jboss/set/draw/LotteryDrawingTest.java [hunk 6]
```diff
@@ -222,7 +252,8 @@ public void testAppendingRepositoryForUnsubcription() throws Exception {
         List<Mail> sent = mailbox.getMailsSentTo(email);
         assertEquals(1, sent.size());
         // Check directly for this URL as we know it is valid one. Using templates might result in a corrupted URL overall
-        assertTrue(sent.get(0).getText().contains("https://github.com/The-Huginn/jira-issue-lottery/blob/main/.github/jira-issue-lottery.yml"));
```
Comment (The-Huginn) [inline src/test/java/org/jboss/set/draw/LotteryDrawingTest.java:225]: this doesn't seem right, as this is a single test with concrete value, with the intention of making sure a valid link is generated

**Review by The-Huginn — CHANGES_REQUESTED** [review]

File: src/main/resources/application.properties (hunk 5 above)
Comment (xjusko) [reply src/main/resources/application.properties:3]: Should I create a new issue for that? Then I can add a new commit to this PR and mention both issues.

**Comment by xjusko** [comment]
resolves #49

**Pull request closed without merging**